WalletWasabi/BestEffortEndpointConnector.cs

```diff
@@ -0,0 +1,78 @@
+using NBitcoin.Protocol;
+using NBitcoin.Protocol.Behaviors;
+using NBitcoin.Protocol.Connectors;
+using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WalletWasabi
+{
+	public class BestEffortEndpointConnector : IEnpointConnector
+	{
+        private enum ConnectionMode
+        {
+            OnionServiceOnly,
+            AllowGoingThroughTorExitNodes,
+            ClearNet
+        }
+
+        private class EffortState
+        {
+            public  ConnectionMode Mode { get; set; }
+            public  DateTimeOffset LastConnectionTime { get; set; }
+            public EffortState(ConnectionMode mode, DateTimeOffset lastConnectionTime)
+            {
+                Mode = mode;
+                LastConnectionTime = lastConnectionTime;
+            }
+        }
+
+        private DefaultEndpointConnector _connector;
+        private EffortState _state;
+
+        public BestEffortEndpointConnector()
+            : this(
+                new DefaultEndpointConnector(allowOnlyTorEndpoints: true), 
+                new EffortState(ConnectionMode.OnionServiceOnly, DateTimeOffset.UtcNow))
+        {
+        }
+
+        private BestEffortEndpointConnector(DefaultEndpointConnector connector, EffortState state)
+        {
+            _connector = connector;
+            _state = state;
+        }
+
+		public IEnpointConnector Clone()
+		{
+			return new BestEffortEndpointConnector((DefaultEndpointConnector)_connector.Clone(), _state);
+		}
+
+		public async Task ConnectSocket(Socket socket, EndPoint endPoint, NodeConnectionParameters nodeConnectionParameters, CancellationToken cancellationToken)
+		{
+            var elapsedTimeSinceLastConnection = DateTimeOffset.UtcNow - _state.LastConnectionTime;
+            var nodesGroup = NodesGroup.GetNodeGroup(nodeConnectionParameters);
```

Comment:
If I understand this correctly, then this call is equivalent to `var nodesGroup = NodesGroup.GetNodeGroup(nodeConnectionParameters.TemplateBehaviors);`. Then this `GetNodeGroup` method gives you a `NodeGroup` from the first behavior. That sounds a little brittle as it is dependent on ordering of those behaviors.

`GetNodeGroup` may return `null` if there are no `NodesGroupBehavior` defined. Yet, that probably cannot happen because Wasabi seems to use `TrustedP2pBehavior` for peers. Do we use `TrustedP2pBehavior` or something else?

I'm curious, can we get multiple different `nodesGroup`s? If so, then we would have one EffortState for multiple `nodeGroup`s.

Reply:
We use only one NodesGroup and it makes no sense to have more than one, at least for us. We could pass the NodesGroup as a parameter or subscribe to some event to know if we are connected to nodes or not.

Reply:
Great.

> We use only one NodesGroup and it makes no sense to have more than one, at least for us.

My concern here is that it may break when somebody works on something else (even when now it makes no sense). It's far from obvious.

Reply:
I changed this code.